### PR TITLE
Use Util.Persistent to reimplemented promoted files store

### LINKED ISSUE #916: Change the format of all persistent files
Currently dune stores various files in `_build` in S-expression format. These are not versioned which means that when we change their format Dune just crashes. In particular we changed the format of a few files for Dune 1.0.0 and running dune on a workspace where there is already a `_build` directory created by jbuilder result in a crash.

We recently introduced `Utils.Persistent` for performance reasons. However it also handles versioning correctly, so we should just use it for everything. To make debugging easier, we should add a `val to_sexp : t -> Sexp.t` function to `Utils.Persistent_desc` and add a sub-command `dune dump <file>` that would dump the contents of the file in S-expression format.

The following places in the code needs updating to use `Utils.Persistent_desc`:

- `Build_system.Promoted_to_delete`
- `Action.Promotion`
- `Vfile_kind`

### PR DESCRIPTION
**Pull request by rgrinberg**
Fix #916 

Will add the other files as separate commits

### CODE REVIEW

**Comment by rgrinberg** [comment]
@diml this is ready for review

File: src/utils.ml [hunk 1]
```diff
@@ -155,6 +155,9 @@ end
 module Persistent(D : Persistent_desc) = struct
   let magic = sprintf "DUNE-%sv%d:" D.name D.version
 
+  let to_out_string (v : D.t) =
+    Marshal.to_string v []
```
Comment (diml) [inline src/utils.ml]: Shouldn't this contain `magic` as well?

File: src/vfile_kind.mli [hunk 2]
```diff
@@ -12,20 +12,16 @@ module type S = sig
...
-    (F : functor (C : Sexp.Combinators) -> sig val t : T.t C.t end)
-  : S with type t = T.t
-
-module Make_full
-    (T : sig type t end)
-    (To_sexp : sig val t : Path.t -> T.t -> Sexp.t end)
-    (Of_sexp : sig val t : Path.t -> Sexp.Ast.t -> T.t end)
+    (T : sig
+       type t
+       val t : t Sexp.To_sexp.t
```
Comment (diml) [inline src/vfile_kind.mli]: There should be a `name` in here as well, otherwise we won't be able to distinguish the various files for `dune dump`.

**Comment by rgrinberg** [comment]
@diml have another look

**Pull request merged**